flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionLookup.java

```diff
@@ -16,19 +16,44 @@
  * limitations under the License.
  */
 
-package org.apache.flink.table.expressions.catalog;
+package org.apache.flink.table.catalog;
 
 import org.apache.flink.annotation.Internal;
 import org.apache.flink.table.functions.FunctionDefinition;
 
+import java.util.Optional;
+
 /**
- * Catalog of functions that can resolve the name of a function to a {@link FunctionDefinition}.
+ * Catalog of functions that can resolve the name of a function to a {@link FunctionLookup.Result}.
  */
 @Internal
-public interface FunctionDefinitionCatalog {
+public interface FunctionLookup {
```

Comment:
I know it will be long but maybe `FunctionDefintionLookup`. It would be cleaner that we lookup only the defintion and we e.g. do not createthe implementation.

Reply:
Yes, but we also return the object identifier. So it is more than just the definition. The catalog is also called `FunctionCatalog` and not `FunctionDefinitionCatalog`.

Reply:
True, but we do return the object identifier, but it is still closer to `FunctionDefinition` than to e.g. Scalar**Function**, Aggregate**Function** etc. But that was a minor suggestion. I am absolutely fine with that name.